### PR TITLE
[WONT MERGE] POC Prototype checkpointing without re-entrant autograd

### LINKED ISSUE #62654: Checkpointing using saved tensors default hooks
## 🚀 Feature
Checkpointing using saved tensors default hooks (no re-entrant autograd!)

## Motivation

[Checkpointing](https://pytorch.org/docs/stable/checkpoint.html?highlight=checkpoint#torch.utils.checkpoint.checkpoint) allows users to trade off compute for memory by not saving the intermediate results during the forward pass and by recomputing the forward during the backward pass.

However, the current implementation uses re-entrant autograd (it calls backward from inside the backward).

I'd like to explore an alternative to implement checkpointing using the [newly introduced saved tensors default hooks](https://github.com/pytorch/pytorch/pull/62563).

## Pitch

```python
class Checkpoint():
    def __init__(self, function):
        self.function = function
        
    def __call__(self, *args):
        storage = []
        counter = 0
        
        def pack(x):
            nonlocal counter, storage
            # we could store something else than counter, but not None because that's a valid saved tensor value
            storage.append(counter) 
            counter += 1
            return counter - 1
        
        def unpack(x):
            nonlocal counter, storage
            if isinstance(storage[x], int):
                counter_inner = 0
                def inner_pack(inner):
                    nonlocal counter, storage, counter_inner
                    storage[counter_inner] = inner
                    counter_inner += 1
                    return None
                
                torch.autograd.graph.set_saved_tensors_default_hooks(inner_pack, lambda x: None)
                with torch.set_grad_enabled(True):
                    y = self.function(*args)
                torch.autograd.graph.reset_saved_tensors_default_hooks()
                
            assert not isinstance(storage[x], int)
            return storage[x]
        
        torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
        y = self.function(*args)
        torch.autograd.graph.reset_saved_tensors_default_hooks()
        return y


model = Checkpoint(ModelBase())
head = ModelHead()
head(model(x)).backward()
```

## Alternatives

The current Checkpointing implementation, which uses re-entrant autograd.

## Additional context

Limitations with the code in the pitch:
* Does not handle storing/restoring RNG state
* Assumes that intermediate values will be saved in the same order during both calls to `self.function`
* Conflicts with other usages of saved tensors default hooks
* Not thread-safe (for now)

### PR DESCRIPTION
**Pull request by Varal7**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63524
* #63425
* __->__ #62964

Fixes #62654

Assumes that intermediate values will be saved in the same order during both calls to `self.function`

Differential Revision: [D30191052](https://our.internmc.facebook.com/intern/diff/D30191052)

### CODE REVIEW

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30191052).

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -275,3 +275,81 @@ def forward(input):
...
+            first input as ``activation`` and the second input as ``hidden``
+        preserve_rng_state(bool, optional, default=True):  Omit stashing and restoring
+            the RNG state during each checkpoint.
+    """
+    def __init__(self, function, preserve_rng_state=True):
+        self.function = function
+        self.preserve_rng_state = preserve_rng_state
+
+    def __call__(self, *args):
+        check_backward_validity(*args)
```
Comment (albanD) [inline torch/utils/checkpoint.py:297]: This is not necessary anymore!!! :)

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -275,3 +275,81 @@ def forward(input):
...
+        self.had_autocast_in_fwd = torch.is_autocast_enabled()
+
+        if self.preserve_rng_state:
+            self.fwd_cpu_state = torch.get_rng_state()
+            # Don't eagerly initialize the cuda context by accident.
+            # (If the user intends that the context is initialized later, within their
+            # run_function, we SHOULD actually stash the cuda state here.  Unfortunately,
+            # we have no way to anticipate this will happen before we run the function.)
```
Comment (albanD) [inline torch/utils/checkpoint.py:305]: I think we can actually detect that after the user function ran and raise a nice error about it

File: torch/utils/checkpoint.py [hunk 3]
```diff
@@ -275,3 +275,81 @@ def forward(input):
...
+                if self.preserve_rng_state and self.had_cuda_in_fwd:
+                    rng_devices = self.fwd_gpu_devices
+                with torch.random.fork_rng(devices=rng_devices, enabled=self.preserve_rng_state):
+                    if self.preserve_rng_state:
+                        torch.set_rng_state(self.fwd_cpu_state)
+                        if self.had_cuda_in_fwd:
+                            set_device_states(self.fwd_gpu_devices, self.fwd_gpu_states)
+                    #  detached_inputs = detach_variable(tuple(inputs))
+                    with torch.enable_grad(), torch.cuda.amp.autocast(self.had_autocast_in_fwd):
+                        torch.autograd.graph.set_saved_tensors_default_hooks(inner_pack, lambda x: None)
```
Comment (albanD) [inline torch/utils/checkpoint.py:345]: What would that unpack lead to? Should we just make it a function that raise an error for the (very unlikely) case someone manages to trigger it?

File: torch/utils/checkpoint.py [hunk 4]
```diff
@@ -275,3 +275,81 @@ def forward(input):
         input = checkpoint(run_function(start, end, functions), input,
                            preserve_rng_state=preserve)
     return run_function(end + 1, len(functions) - 1, functions)(input)
+
+
+class Checkpoint():
```
Comment (albanD) [inline torch/utils/checkpoint.py:280]: Make this an nn.Module so that it works well when containing a Module

**Review by albanD — COMMENTED** [review]
Looks quite nifty!
I think we want to double check a couple things:
- With torchviz that we generate the proper graph
- Meaning that autograd.grad should work just fine
- And hooks on the parameters are triggered only once, even if they are used inside and outside the checkpoint
- And that should work with DDP!

**Comment by Varal7** [comment]
> * And hooks on the parameters are triggered only once, even if they are used inside and outside the checkpoint

Not sure what you mean by that.
Is it something like this where `a` is saved both inside and outside the checkpoint?
```
class Square(torch.nn.Module):
    def forward(self, x):
        return torch.pow(x, 2)

a = torch.ones(2, 2, requires_grad=True)
b = Checkpoint(Square())(a).sum()
c = Square()(a).sum()
d = b + c
```

**Comment by albanD** [comment]
@Varal7 
What I mean is that the following code should print "hey" only once. But with the existing checkpoint it prints it twice:
```
import torch
from torch.nn import Linear
from torch.utils import checkpoint

mod = Linear(10, 10)
def hook(grad):
    print("hey")
mod.weight.register_hook(hook)

inp = torch.rand(1, 10, requires_grad=True)

inter = mod(inp)
out = checkpoint.checkpoint(mod, inter)

out.sum().backward()
```

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30191052).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30191052).

File: test/test_autograd.py [hunk 5]
```diff
@@ -4506,6 +4532,57 @@ def test_checkpoint_valid_reset_on_error(self):
...
+
+    def test_checkpointing_with_hooks_dataparallel(self):
+        a = torch.randn(20, 20, requires_grad=True)
+
+        b = torch.exp(a).sum()
+        b.backward()
+        b_grad = a.grad
+
+        a.grad = None
+        c = torch.nn.DataParallel(Checkpoint(torch.exp))(a).sum()
```
Comment (Varal7) [inline test/test_autograd.py:4562]: This currently fails but should work after #62909 is merged.

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30191052).

File: test/test_autograd.py [hunk 6]
```diff
@@ -4506,6 +4532,57 @@ def test_checkpoint_valid_reset_on_error(self):
...
+        def hook(grad):
+            nonlocal count
+            count += 1
+
+        w.register_hook(hook)
+        x = torch.rand(10, 10, requires_grad=True)
+        h = w * x  # Using w outisde the checkpoint
+        out = Checkpoint(lambda x: w * x)(h)  # Using w inside the checkpoint
+
+        warnings.simplefilter('always')
```
Comment (Varal7) [inline test/test_autograd.py:4581]: woops

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30191052).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30191052).

**Comment by ezyang** [comment]
Does the new checkpointing implementation support torch.autograd.grad calls as well?

**Comment by albanD** [comment]
Yes it does.

By the way https://github.com/pytorch/pytorch/issues/65537 is tracking the work that is planned to merge this feature.
This stack is a POC and won't be merged as-is.

**Pull request closed without merging**